### ISSUE TITLE
Parallel proof exploration plan (draft)

### ISSUE THREAD

**Issue by nwatson22** [runtimeverification/kontrol#131]
## In `pyk`:
1. Make `Proof` disk access functions (`read_proof_data()` and `write_proof_data()`) safe to be used across multiple processes by adding a parameter of type `multiprocessing.Lock` which would be obtained before reading/writing and released after.

## In `kontrol`:
1. Remove the current parallel proof exploration with `ProcessPool.map()`
2. Add 2 `multiprocessing.Queue`s. One is the in queue and the other is the out queue. The in queue will contain different types of named tuples for different types of jobs:
  a. `(proof_id: str, node_id: int, lock: multiprocessing.Lock)` for advancing one node in the proof 
  b. `(proof_id, lock)` for initializing a proof (essentially just calling `method_to_apr_proof()`)
These jobs will submit outputs to the out queue of one of these forms:
  a. `(proof_id, node_id, result: ExtendResult)` to let the main process know which proof to extend with where and what.
  b. `(proof_id)` for letting the main process know that the proof is done being initialized and its execution can be queued.
  c. `(proof_id, node_id)` for letting the main process know whether the node can be subsumed so it can modify the KCFG accordingly.
3. Create a map that stores, for each test, a `multiprocessing.Lock` that will be used to lock disk access to the corresponding proof
4. Create a map to cache the `Proof` objects (for the main process only)
5. Create and start the requested number of worker processes
6. Submit a type (b) job to the in queue for each proof being run.

Repeat the following in the main process:

1. Pull from the out queue.
2. Depending on the type of output:
  a. 
    i. Load the proof from the cache by its ID.
    ii. Call `extend_kcfg()` with the produced `ExtendResult` and the appropriate
    iii. Add the new nodes created by the extension as type (a) jobs
  b.
    i. Load the proof from disk based on its ID. Cache it in the proofs map.
    ii. Add the proof's init node as a type (a) job.
  c.
    i. Load the proof from the cache by its ID
    ii. Add the subsumption relation to the KCFG
3. Perform in-between steps which already exist:
  i. Check if iteration bound for the specific proof has been reached and cut off execution of that proof if necessary.
  ii. Check if any node has failed and the `fail_fast` flag is on and cut off execution of that proof if necessary.
  iii. Check if any nodes need to be marked as terminal based on the `KCFGSemantics` and mark them if necessary (_check_all_terminals).
  iv. If the proof has failed, call `save_failure_info()` on the `Prover`.
  v. Save the proof to disk

Repeat the following in each worker process:

1. Pull from the in queue.
2. Depending on the type of output:
  a.
    i. Read proof from disk. Get `CTerm` at the specified node and the target node.
    ii. Check implication between specified node and target node. If it is able to be subsumed, submit a type (c) result to the out queue.
    ii. Otherwise, call `extend_cterm()` on this `CTerm` to get the `ExtendResult` and submit this to the out queue (type (a) result).
  b.
    i. Create proof from method with `method_to_apr_proof()`
    ii. Save proof to disk.
    iii. Submit a type (b) result to the out queue.
    
- Need some way to share ProveOptions to the worker threads
- Keep support for EqualityProof and APRBMCProof
- Proofs in each worker need to reread from disk every time because the proof could have changed. Main process doesn't have to because it is the only one that changes the `Proof`s after initialization
- How to know when to stop pulling from get queue (count jobs submitted and jobs received)
- Make sure worker processes are cleaned up
- How to share KoreServers

**Comment by tothtamas28** [comment]
Although taking care of initialization and exploration of proofs in a single loop is optimal with respect to resource utilization, I'd still split these into two separate flows for now. This has the following advantages:
* Two simpler processes instead of a single complex one.
* We can have a generic multi-threaded proof process in `pyk` (along with supporting tools for debugging, etc.), and semantic-specific initialization processes in downstream projects.

For now, we should focus on parallel exploration. Later, once we're confident in our changes, we can figure out a way to build initialization hooks into the generic proof process.

**Comment by ehildenb** [comment]
A couple questions:

- Step 5 is describing how to extend the KCFG. I think that `KCFGExplore.extend_kcfg` already does this? Could this step be simplified to say "Pull from the `out_queue` the `ExtendResult`, read the proof from disk, use `KCFGExplore.extend_kcfg` to update the proof, then write the proof to disk."?
- It would be nice if it was clearer in the description what the main process and the worker processes will be doing. Can we have the description updated to clarify in separate sections):
  - Which data is needed (`in_queue`, `out_queue`, etc...)
  - What the main process is doing (pulling items from the `out_queue`, processing them, adding items to the `in_queue`).
  - What the worker processes are doing (pulling items from the `in_queue`, processing them, adding `ExtendResult`s to the `out_queue`).

**Comment by nwatson22** [comment]
> Step 5 is describing how to extend the KCFG. I think that `KCFGExplore.extend_kcfg` already does this? Could this step be simplified to say "Pull from the `out_queue` the `ExtendResult`, read the proof from disk, use `KCFGExplore.extend_kcfg` to update the proof, then write the proof to disk."?

@ehildenb `extend_kcfg` only adds the next node(s) and appropriate edge type(s). The steps now listed under 4.d in the main process execution loop are currently done in `APRProver.advance_proof()`. I included them here because `advance_proof()` also calls `advance_pending_node()`, which is the `extend_cterm` step, which we want to do in the worker threads, and the `extend_kcfg` step, together. So we would be taking these functions from there and moving them to a separate function or functions which are called by the main loop in the main process.

> It would be nice if it was clearer in the description what the main process and the worker processes will be doing. Can we have the description updated to clarify in separate sections): 

I've updated now to be clearer about the data items needed and the execution loops for both main and worker processes.

**Comment by tothtamas28** [comment]
Let's discuss this in more abstract terms. Consider

```python3
class Proof:
    def status(self) -> ProofStatus:
        ...
```

and

```python3
class Prover(Generic[P, S, U]):
    def steps(self, proof: P) -> Iterable[S]:
        ...

    def advance(self, proof: P, step: S) -> U:
        ...

    # Should return P to be more flexible, but let's assume this for implicity
    def commit(self, proof: P, update: U) -> None:
        ...
```

Here, `Prover` is responsible for making atomic updates on a `Proof`. The contract between `Proof` and `Prover` is based on `Proof.status`, in particular
```
IF proof.status() is ProofStatus.PENDING THEN list(prover.steps(proof)) != []
```

I think we have little work to do to implement such an interface for all our currently supported proof types.

A simple proof loop could then look like this (in Python-like pseudocode):

```python3
def prove_parallel(
    proofs: list[Proof],
    # We need a way to map proofs to provers, but for simplicity, I'll assume it as a given
    provers: dict[Proof, Prover],
):
    pending: dict[Future[Update], Proof] = {}

    def submit(proof: Proof, pool: Executor) -> None:
        prover = provers[proof]
        for step in prover.steps(proof):  # <-- get next steps (represented by e.g. pending nodes, ...)
            future = pool.submit(prover.advance, step)  # <-- schedule steps for execution
            pending[future] = proof

    with ProcessPoolExecutor() as pool:
        for proof in proofs:
            submit(proof, pool)

        while pending:
            future = pool.wait(pending)
            proof = pending[future]
            prover = provers[proof]
            update = future.result()
            prover.commit(proof, update)  # <-- update the proof (can be in-memory, access disk with locking, ...)

            match proof.status():
                # terminate on first failure, yield partial results, etc.
                ...

            submit(proof, pool)
            pending.pop(future)
```

This is pretty much a bare bones solution, but it can be extended with e.g. dependencies between proofs (i.e. based on the call graph), scheduling of steps, etc.

@nwatson22, do you think this model is flexible enough to accommodate your implementation plan, or there's something missing that should be considered?

If we can agree on an interface, we can implement toy provers for testing the exploration logic (e.g. discover a predefined tree of `int`-s stored in a `dict`) before writing the actual provers.

**Comment by nwatson22** [comment]
@tothtamas28 I think the interface proposed looks good. I am curious about how submitting the `prover.advance` function to the process pool works. We have additional information that is passed to `cterm_execute`, such as the execute depth, module name, cut point rules list, and terminal rules list. I'm wondering if we can have them as members. 3 of them should be set and the beginning and never change, but currently `APRProver` is computing the `module_name` dynamically with
```
        module_name = self.circularities_module_name if self.nonzero_depth(node) else self.dependencies_module_name
```

I opened a PR https://github.com/runtimeverification/pyk/pull/709 that adds the interface to the `Prover` and `Proof` classes and tests it with an implementation of `prove_parallel`.

Currently the problem I'm having with that is that if it finishes a job and gets to the `submit(proof, pool)` part of the main loop, it submits all `prover.steps(proof)`, which may include steps which have already been started but not yet finished. This results in those nodes being re-explored. I got around this by keeping track of the nodes already started inside of `prove_parallel`:
```
        for step in prover.steps(proof):  # <-- get next steps (represented by e.g. pending nodes, ...)
            if step in explored:
                continue
            explored.add(step)
            future = pool.submit(prover.advance, proof, step)  # <-- schedule steps for execution
            pending[future] = proof
```

**Comment by tothtamas28** [comment]
> I'm wondering if we can have them as members of `APRProver`.

Although the name is misleading the `Prover` interface as proposed is not intended to be a replacement for the `Prover` class hierarchy  that `APRProver` belongs to, at least not in the first iteration. Instead, consider it an adapter between classes that manipulate a `Proof` (like `APRProver`) and the function `prove_parallel`. So it can wrap an `APRProver` or any other class necessary to make progress on the proof.

> Currently the problem I'm having with that is that if it finishes a job and gets to the `submit(proof, pool)` part of the main loop, it submits all `prover.steps(proof)`, which may include steps which have already been started but not yet finished.

Good point. We have a few options:

* Change the `Prover` interface to return some `initial_steps` and the rest in `commit`.
* Make it part of `prove_parallel` state, i.e. what you propose.
* Make it part of `Prover` state. This is also fine as the lifecycle of a `Prover` does not have to extend beyond a call to `prove_parallel`, but it makes correct implementation of a `Prover` a bit more complex.

Whatever the design, we should think about what implications it has on the contract as that will dictate what abstract properties we should test for. Please think this through.

**Comment by nwatson22** [comment]
I'm trying out the `initial_steps` idea. To avoid having to store the begun nodes in the external state, now we have `commit` return only those nodes that were unlocked by the commit. As you pointed out that modifies the contract as you described it. Now it's

**Comment by tothtamas28** [comment]
> I think this would be useful for implementing `fail_fast` functionality

Fail-fast should be implemented in `prove_parallel`: you shut down as soon as `proof.status() is ProofStatus.FAILED` for some `proof`.

> I think it would require adding another method to the Prover interface like `begin_step()` (since advance() is being called on a separate process) 

Not necessarily, it can be part of the contract that a step will only be returned once.

> For the contract

It imposes the requirement a step has to be `Hashable`. Also, a step has to be returned until it's result is committed (this enables the stale step removal feature), but not after (to guarantee progress).

Regarding step removal, I'm not sure it should be the responsibility of the `Prover`. But at least, `Prover` should not assume that the results of a step later withdrawn will not be committed. We can figure out the rest later.

But it looks like you lean towards the "store state in `prove_parallel`" option. Please write up the contract between, `Proof`, `Prover` and `prove_parallel`. After that, we will attempt to turn that into code (documentation, assertions, test cases, etc).

**Comment by nwatson22** [comment]
@tothtamas28 
There are a few things I'm still not sure about.
- we are passing objects implementing `Proof` to `pool.submit()`, where the idea I had written down was to pass proof IDs and then read the proof  from disk in `prover.advance()` based on that ID. Is `Proof` meant to be more of an identifier for actual proofs that can also get the status of the proof it identifies? Should we add something like `id(self) -> str` to the `Proof` interface and then have `prover.advance()` take those IDs instead? 
- I'm not sure how to be more specific about requirements relating to guaranteeing progress on a proof, since for, e.g., `APRProof`s, there's no guarantee they eventually stop `PENDING` no matter how many times we advance and commit new steps.

Would this be roughly what you mean by the contract?

## Proof:
should represent a computer proof of a single claim.

### Must implement methods:

`status(self) -> ProofStatus`, which returns one of `ProofStatus.PASSED`, `ProofStatus.FAILING`, or `ProofStatus.PENDING`

- `proof.status() == ProofStatus.PASSED` should mean that the claim that `proof` is trying to prove has been proven.
- `proof.status() == ProofStatus.FAILED` should mean that the claim that `proof` is trying to prove has not been proven, but the proof cannot advance further.
- `proof.status() == ProofStatus.PENDING` should mean that the claim that `proof` is trying to prove has not yet been proven, and the proof can advance further so might still be proven with more calls to `prover.commit(proof, update)`
- If `proof.status() == ProofStatus.PENDING`, `prover.steps(proof)` must be nonempty.

## Prover:
`Prover` should be used to store any objects and other data needed to make progress on `P`s. may be specific to a single `P` or able to be used by multiple `P`s.

a single `P` must only ever interact through this interface with one prover object.


### Parameterized by:

`P`
- must implement `Proof` interface
- should be small, since it is intended to be passed between threads?

`S`
- must be hashable
- must be immutable
- should be small, since it is intended to be passed between threads?
- should tell `prover.advance()` which  computation to perform.

`U`
- must be immutable
- should be small, since it is intended to be passed between threads?
- should represent instructions for making a specific small change to the state of a `P` based on the results of a computation specified by an `S`

### Must implement methods:

`steps(self, proof: P) -> Iterable[step]`

- must not modify `self` or `proof`.
- should return a list of `S` which represents all the atomic computation jobs of `proof` that have not yet been completed and submitted, and are available given the current state of `proof`.
- for all steps `step`, after `prover.commit(proof, prover.advance(proof, step))` is called, `step`, should never again be returned by `prover.step(proof)`.

`advance(self, proof: P, step: S) -> U`

- must not modify `self` or `proof`.
- may be nondeterministic.
- should perform some nontrivial computation needed to advance the proof, which can be done independently of calls to `advance()`. it should return a piece of data that specifies how the results of that computation should update the state of `proof`.
- must only be called on a `step` that was returned by `prover.steps(proof)`.

`commit(self, proof: P, update: U) -> None`

- should update `proof` according to `update`.
- must only be called on an `update` that was returned by `prover.advance(proof, step)` for some `step`.
- after a step `step` becomes available through `prover.steps(proof)`, it can be commmitted via
  `prover.commit(proof, prover.advance(proof, step))` at any time during the lifecycle of `prover`, including
  more than once, and different results could be committed in any order.

**Comment by tothtamas28** [comment]
> we are passing objects implementing `Proof` to `pool.submit()`, where the idea I had written down was to pass proof IDs and then read the proof from disk in `prover.advance()` based on that ID

This design doesn't prohibit such an implementation. Sure, you can have a proof that only stores an `id: str` and performs disk access to implement `status()`.

> Im not sure how to be more specific about requirements relating to guaranteeing progress on a proof, since for, e.g., `APRProofs`, there's no guarantee they eventually stop `PENDING`.

Of course, we cannot expect all provers to be sound an complete. With progress I mean if `advance(s) = u` for some step `s`, and then you `commit(u)`, then `s in steps()` will no longer hold.

**Comment by tothtamas28** [comment]
> P should be small, since it is intended to be passed between threads?

Good point. Maybe `advance` should not receive a `P`?

**Comment by tothtamas28** [comment]
Here's my list of properties (some of them are also present in your list):

* `step` is `Hashable`.
* `step` and `update` are pickleable.
* If `proof.status() is ProofStatus.PENDING`, then `list(prover.steps(proof)) != []`.
* If `proof.status() is ProofStatus.PASSED`, then `list(prover.steps(proof)) == []`.
* If `proof.status() is ProofStatus.PASSED`, then it remains so even after some `prover.commit(proof, u)`.
* If `proof.status() is ProofStatus.FAILED`, then it remains so even after some `prover.commit(proof, u)`.
* If `prover.advance(s) == u`, then after `prover.commit(proof, u)`, no longer `s in prover.steps(proof)`.
* If `s in prover.steps(proof)` at some point and `prover.advance(s) == u`, then `prover.commit(proof, u)` may be called.
* `proof.status()` may only change due to some `prover.commit(proof, u)`.

Can you think of additional properties of this style?

**Comment by nwatson22** [comment]
@tothtamas28 I've added your suggestions and changed a few things. `advance` is now static to avoid having to pickle the entire `Prover` every time we start a new task (assuming I've understood this correctly). For `APRProof` only a few `Prover` details are needed: `execute_depth`, `cut_point_rules`, and `terminal_rules`. We could pass these as part of an `S`, although ideally we'd want to have the processes cache them. Perhaps that doesen't need to be explicitly handled in this contract.

I also removed the `proof` parameter from `advance`, which for `APRProofs` should leave open either the possibility of passing in a proof identifier as part of `step` which `advance` would have to use to read from disk, as described above or just passing in a `CTerm` to extend as part of `step`.

# Contract:

## Proof:
should represent a computer proof of a single claim.

### Must implement methods:

`status(self) -> ProofStatus`,

- must return one of `ProofStatus.PASSED`, `ProofStatus.FAILING`, or `ProofStatus.PENDING`
- `proof.status() is ProofStatus.PASSED` should mean that the claim that `proof` is trying to prove has been proven.
- `proof.status() is ProofStatus.FAILED` should mean that the claim that `proof` is trying to prove has not been proven, but the proof cannot advance further.
- `proof.status() is ProofStatus.PENDING` should mean that the claim that `proof` is trying to prove has not yet been proven, and the proof can advance further so might still be proven with more calls to `prover.commit(proof, update)`
- if `proof.status() is ProofStatus.PENDING`, `prover.steps(proof)` must be nonempty.
- if `proof.status() is ProofStatus.PASSED`, `prover.steps(proof)` must be empty.
- once `proof.status() is ProofStatus.PASSED or ProofStatus.FAILED`, it must remain so.

## Prover:
`Prover` should be used to store any objects and other data needed to make progress on `P`s. may be specific to a single `P` or able to be used by multiple `P`s.

a single `P` must only ever interact through this interface with one prover object.


### Parameterized by:

`P`
- must implement `Proof` interface

`S`
- must be hashable
- must be immutable
- must be picklable
- should be small, since it is intended to be passed between threads
- should tell `Prover.advance()` which  computation to perform.

`U`
- must be immutable
- must be picklable
- should be small, since it is intended to be passed between threads
- should represent instructions for making a specific small change to the state of a `P` based on the results of a computation specified by an `S`

### Must implement methods:

`steps(self, proof: P) -> Iterable[step]`

- must not modify `self` or `proof`.
- should return a list of `S` which represents all the atomic computation jobs of `proof` that have not yet been completed and submitted, and are available given the current state of `proof`.
- for all steps `step`, after `prover.commit(proof, Prover.advance(step))` is called, `step`, must never again be returned by `prover.step(proof)`.
- output steps must not change except with calls to `prover.commit(proof, update)` for some `update`.

`advance(step: S) -> U`

- static method.
- may be nondeterministic.
- should perform some nontrivial computation needed to advance a proof, which can be done independently of other calls to `advance()`. it should return a piece of data that specifies how the results of that computation should update the state of a `P`.
- Able to be called on any step returned by `prover.steps(proof)`

`commit(self, proof: P, update: U) -> None`

- should update `proof` according to `update`.
- must only be called on an `update` that was returned by `Prover.advance(step)` for some `step`, where .`step` was returned by `prover.steps(proof)`.
- No other method may cause a change to `proof.status()` besides `prover.commit(proof, update)`.
- after a step `step` becomes available through `prover.steps(proof)`, it may be commmitted via `prover.commit(proof, Prover.advance(step))` at any time during the lifecycle of `prover`, including more than once, and different results could be committed in any order.

**Comment by tothtamas28** [comment]
> `advance` is now static to avoid having to pickle the entire `Prover` every time we start a new task

Good point. Instead of making it a `staticmethod`, consider:

```python3
class Prover(Generic[P, U]):
    def steps(self, proof: P) -> Iterable[ProofStep[U]]:
        ...

    def commit(self, proof: P, update: U) -> None:
        ...

class ProofStep(Generic[U]):
    def execute(self) -> U:
        ...
```

> would it be a good idea to add a method to `Prover` to extract this information

The overhead of pickling objects is probably negligible compared to the time it takes to do the symbolic execution, so for now I'd avoid introducing complexity to reduce it. If it turns out to be an issue, we can figure out a way to share data between tasks running on the same process.

**Comment by tothtamas28** [comment]
Notes regarding a few interrelated extensions that we should implement at some point.

## Exception handling

Exceptions can occur during `step.exec()`. We have a few ways of handling this.

1. Fail (current design)
2. No longer use the prover that emitted the failing step, continue with other provers
3. Retry

## Resource management

`step.exec` might require certain resources. These can be owned by

1. The step
2. The proof the step is performed on
3. The prover that emitted the step

Case 1 can be handled within `exec`.

Case 3 can be handled outside of `prove_parallel`, although if we want to implement more refined exception handling (Cases 2 and 3 there), we need methods `prove_parallel` can call to notify the prover to acquire / release its resources.

Case 2 also requires methods `prove_parallel` can call to notify the prover.

## Subproofs

Requires a parameter `deps: Mapping[str, str]` on `prove_parallel`. This will be checked for cycles: if there is one, an exception is raised.

We also need some way to notify the prover about passing subproofs. This can be conveniently done as part of resource management (Case 2 there):

```python3
def ready(proof: P, subproofs: list[Proof]) -> None:
    ...
```

(A `TypeVar` can also be introduced for subproofs.)

This would be called when `proof` is ready to run, i.e. all dependencies are passing. In case of a `kore-rpc`-based prover, the prover would then start a sever for the proof, and call the `add-module` endpoint to add lemmas that correspond to the subproofs.

**Comment by nwatson22** [comment]
# Proposed addition to contract:
## ProcessData
- should represent data that is specific to a single worker process.
- must be picklable.
- may store data per proof or prover, e.g. in a map.

### Must implement methods:

`init(self) -> None`

- called at the beginning of execution of each worker process
- should be used to initialize resources stored as members of the `ProcessData`

`cleanup(self) -> None`

- called at the end of execution of each worker process
- should be used to clean up resources stored as members of the `ProcessData`

# Resource management

For APRProof implementation:
Needed for `step.exec`:

Owned by steps:
- prover ID (to look up the prover x process data)
- cterm: cterm to step from
- node_id: must be returned back with the result to tell which node to apply the ExtendResult to.
- target_cterm: CTerm to check subsumption into
- is_terminal and target_is_terminal: parameters to determine whether or not to check subsumption and to return a 'subsumption failed' result
- execute_depth, terminal_rules, cut_point_rules - `extend_cterm()` parameters

Owned by the processes (`ProcessData`), one copy each
- kore_server: store in map in APRProofProcessData. When a new proof/prover ID is received, initialize a new server for it.

global options (Ideally avoid pickling, but alternatively pass them to worker process once as part of ProofData):
- module_name: kprint.main_module
- bug_report: global parameter into KoreClient()
- trace_rewrites: global option for KCFGExplore()
- smt_timeout: used for kore_server initialization
- smt_retry_limit: used for kore_server initialization
- smt_tactic: used for kore_server initialization
- haskell_log_format: used for kore_server initialization
- haskell_log_entries: used for kore_server initialization
- log_axioms_file: used for kore_server initialization
- kprint: used for KCFGExplore initialization
- kcfg_semantics: used for KCFGExplore initialization
- definition_dir: used for kore_server initialization
- command: used for kore_server initialization

For APRProvers, we assume proof/prover is 1 to 1. Inside exec, we check to see if the server for a given
proof/prover ID exists in `kore_server` map, and start it if not. Then, based on the port of that server, we
start up a `KoreClient` and `KCFGExplore` using that client. We then use the parameters `is_terminal` and
`target_is_terminal` to determine whether to check terminal subsumption this step. `is_terminal` means the
node that's being extended is terminal, and `target_is_terminal` means the target node of the proof is
terminal. If the target node is not terminal we have to check every time if a node is terminal so we can stop
at that point. If it is terminal, we only have to check subsumption once the reached node is terminal. If the
current node is terminal and fails to subsume, we don't have to try to extend.

Finally after subsumption is (optionally) checked, if no subsumption result has been returned, return the
result of `extend_cterm`.

**Comment by yale-vinson** [comment]
Issue blocked by [Backend Booster #383](https://github.com/runtimeverification/hs-backend-booster/issues/383)

**Comment by nwatson22** [comment]
Closing since an MVP version parallelization has been merged into pyk.
